odoo/http.py

```diff
@@ -418,20 +347,173 @@ def serialize_exception(exception):
+    def from_path(cls, path, filter_ext=('',)):
+        """ Create a :class:`~Stream`: from an addon resource. """
+        path = file_path(path, filter_ext)
+        check = adler32(path.encode("utf-8")) & 0xFFFFFFFF
+        stat = os.stat(path)
+        return cls(
+            type='path',
+            path=path,
+            download_name=os.path.basename(path),
+            etag=f'{int(stat.st_mtime)}-{stat.st_size}-{check}',
+            last_modified=stat.st_mtime,
+            max_age=STATIC_CACHE,
+            size=stat.st_size,
+        )
+
+    @classmethod
+    def from_attachment(cls, attachment):
+        """ Create a :class:`~Stream`: from an attachment. """
+        self = cls(
+            mimetype=attachment.mimetype,
+            download_name=attachment.name,
+            conditional=True,
+            etag=attachment.checksum,
+            max_age=STATIC_CACHE,
+        )
+
+        if attachment.store_fname:
+            self.type = 'path'
+            self.path = werkzeug.security.safe_join(
+                os.path.abspath(config.filestore(request.db)),
+                attachment.store_fname
+            )
+            stat = os.stat(self.path)
+            self.last_modified = stat.st_mtime
+            self.size = stat.st_size
+
+        elif attachment.db_datas:
+            self.type = 'data'
+            self.data = attachment.raw
+            self.last_modified = attachment['__last_update']
+            self.size = len(self.data)
+
+        elif attachment.url:
+            # When the URL targets a file located in an addon, assume it
+            # is a path to the resource. It saves an indirection and
+            # stream the file right away.
+            static_path = root.get_static_file(
+                attachment.url,
+                host=request.httprequest.environ.get('HTTP_HOST', '')
+            )
+            if static_path:
+                self = cls.from_path(static_path)
+            else:
+                self.type = 'url'
+                self.url = attachment.url
+
+        else:
+            self.type = 'data'
+            self.data = b''
+            self.size = 0
+
+        return self
+
+    @classmethod
+    def from_binary_field(cls, record, field):
+        """ Create a :class:`~Stream`: from a binary field. """
+        data_b64 = record[field]
+        self = cls()
+        self.type = 'data'
+        self.data = base64.b64decode(data_b64) if data_b64 else b''
+        self.etag = request.env['ir.attachment']._compute_checksum(self.data)
+        self.last_modified = record['__last_update']  # better than nothing
+        self.max_age = STATIC_CACHE
+        self.size = len(self.data)
+        return self
+
+    def get_response(self, **send_file_kwargs):
+        """
+        Create the corresponding :class:`~Response` for the current stream.
+
+        :param bool as_attachment: Indicate to the browser that it
+            should offer to save the file instead of displaying it.
```

Comment:
Docstring is just wrong.

Reply:
Not that much actually... Out of all the `send_file` arguments, `as_attachment` is the only one that is useful when using the `Stream` API. The others like `mimetype`, `etag`, ... would only override the sensitive values set by stream. I rather not document them.